src/txmempool.cpp

```diff
@@ -187,10 +171,10 @@ bool CTxMemPool::CalculateMemPoolAncestors(const CTxMemPoolEntry &entry, setEntr
         staged_ancestors.erase(stage);
         totalSizeWithAncestors += stageit->GetTxSize();
 
-        if (stageit->GetSizeWithDescendants() + entry.GetTxSize() > limitDescendantSize) {
+        if (stageit->GetSizeWithDescendants() + entry_size > limitDescendantSize) {
             errString = strprintf("exceeds descendant size limit for tx %s [limit: %u]", stageit->GetTx().GetHash().ToString(), limitDescendantSize);
             return false;
-        } else if (stageit->GetCountWithDescendants() + 1 > limitDescendantCount) {
+        } else if (stageit->GetCountWithDescendants() + entry_count > limitDescendantCount) {
             errString = strprintf("too many descendants for tx %s [limit: %u]", stageit->GetTx().GetHash().ToString(), limitDescendantCount);
             return false;
         } else if (totalSizeWithAncestors > limitAncestorSize) {
```

Comment:
I think we have topologies of in-mempool transactions and packages bypassing this limit.

Let's say you have in-mempool, independent Tx_A and Tx_B where their virtual sizes are 30 KvB each.
Let's say you have packages transactions Tx_C and _Tx_D where Tx_C is a child of A and B and Tx_D is a child of Tx_C their virtual sizes are 30 KvB each.

In `CheckPackageLimits`, Tx_C isn't already in the mempool (`mapTx` is only updated in `Finalize`) and as such shouldn't part of `staged_ancestors`. This set should be composed of Tx_A and Tx_B only.

In `CalculateAncestorsAndCheckLimits`, `totalSizeWithAncestors` is the union Tx_A and Tx_B and its sum of 60 KvB is inferior to 101 KvB. After Tx_C and Tx_D are added to the mempool, the chain of transactions ABCD is of the sum 120 KvB.

Further, I don't think it's rejected by the check against `limitDescendantSize`, as Tx_A and Tx_B are evaluated independently, and not as a single cluster as they should be (Tx_A + Tx_C + Tx_D or Tx_B + Tx_C + Tx_D)

Note, you need to replace Tx by chain of transactions because of `MAX_STANDARD_TX_WEIGHT` but I think this reasoning holds ?

Reply:
Thanks for writing the test, I noticed where my reasoning was flawed! 

>  totalSizeWithAncestors is the union Tx_A and Tx_B and its sum of 60 KvB is inferior to 101 KvB. 

We init `totalSizeWithAncestors` with the whole package size from now, instead of the entry only (L164 in `src/txmempool.cpp`). So we have A+B+C+D  and the check against `limitAncestorSize` rejects the package. Maybe the variable could be renamed `packageSizeWithAncestors`, but that's a nit.